### ISSUE TITLE
TypedDict to dict converts to dict[str, object]

### ISSUE THREAD

**Issue by YodaEmbedding** [microsoft/pyright#9541]
**Describe the bug**

`dict()` on a `TypedDict` converts to `dict[str, object]` which causes type issues when doing `len(dict(d)[key])`.

**Code or Screenshots**

Good:

```python
x = object()
len(x)
```

```none
good.py:2:5 - error: Argument of type "object" cannot be assigned to parameter "obj" of type "Sized" in function "len"
    "object" is incompatible with protocol "Sized"
      "__len__" is not present (reportArgumentType)
1 error, 0 warnings, 0 informations
```

Problematic:

```python
from typing import TypedDict

class MyDict(TypedDict):
    k: str

d = MyDict(k="v")
d = dict(d)
d["k2"] = "v2"

len(d["k"])
len(d["k2"])  # No errors here, though.
```

```none
problematic.py:10:5 - error: Argument of type "object" cannot be assigned to parameter "obj" of type "Sized" in function "len"
    "object" is incompatible with protocol "Sized"
      "__len__" is not present (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

**VS Code extension or command-line**
CLI. 1.1.389.

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

A `TypedDict` is a structural type that describes the "shape" of a dictionary. A value that conforms to a TypedDict type is allowed to have additional keys that are not defined in the TypedDict. For this reason, the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/typeddict.html) is clear that type checkers cannot assume that no other keys exist. A TypedDict is therefore a subtype of `Mapping[str, object]`. In fact, you'll see that the class definition for `_TypedDict` in the typeshed `typing.pyi` module derives from `Mapping[str, object]`. 

Not surprisingly, mypy (and other type checkers that conform to the typing spec) generate the same error here as pyright.
